File: activity14 Java/src/com/gdb/domain/SavingsAccount.java
package com.gdb.domain;

import com.gdb.exceptions.*;

public class SavingsAccount extends AbstractAccount {
    private int tenureYears;
    private double minBalance;
    private double interestRate;

    public SavingsAccount(String accountNumber, String name, int age, double balance, String status, String pin, int tenureYears) {
        super(accountNumber, name, age, balance, "SAVINGS", status, pin);
        this.tenureYears = tenureYears;
        this.minBalance = AccountRulesEngine.getSavingsMinBalance(tenureYears);
        this.interestRate = AccountRulesEngine.getSavingsInterestRate(tenureYears);
    }

    public SavingsAccount(String accountNumber, String name, int age, double balance, String status, String pin, double minBalance, double interestRate) {
        super(accountNumber, name, age, balance, "SAVINGS", status, pin);
        this.tenureYears = 0;
        this.minBalance = minBalance;
        this.interestRate = interestRate;
    }

    @Override
    public void processDebit(double amount) throws AccountException {
        if ((this.balance - amount) < this.minBalance) {
            throw new MinimumBalanceViolationException("Cannot breach minimum balance of Rs " + minBalance);
        }
        this.balance -= amount;
    }

    public void applyInterest() {
        double interest = this.balance * (interestRate / 100.0);
        this.balance += interest;
    }

    public int getTenureYears() { return tenureYears; }
    public double getMinBalance() { return minBalance; }
    public double getInterestRate() { return interestRate; }
}

File: activity14 Python/gdb/domain/savings_account.py
# gdb/domain/savings_account.py
from gdb.domain.abstract_account import AbstractAccount
from gdb.exceptions import (
    InactiveAccountException,
    InvalidAmountException,
    MinimumBalanceViolationException
)

class SavingsAccount(AbstractAccount):
    def __init__(self, account_number: str, name: str, age: int, balance: float, status: str = "Active", pin: str = "0000", interest_rate: float = 4.0, minimum_balance: float = 1000.0) -> None:
        super().__init__(account_number, name, age, balance, status, pin)
        self._interest_rate = interest_rate
        self._minimum_balance = minimum_balance

    def calculate_interest(self) -> float:
        return (self._balance * self._interest_rate) / 100.0

    def get_account_type(self) -> str:
        return "Savings"

    def withdraw(self, amount: float) -> None:
        if self._status.lower() != "active":
            raise InactiveAccountException(f"Cannot withdraw from inactive account: {self._account_number}")
        if amount <= 0:
            raise InvalidAmountException(f"Withdrawal amount must be strictly positive: {amount}")
        if (self._balance - amount) < self._minimum_balance:
            raise MinimumBalanceViolationException(f"Withdrawal violates minimum balance requirement of Rs {self._minimum_balance}")
        super().withdraw(amount)
